benchexec/resources.py

```diff
@@ -15,191 +10,157 @@
-        if len(cores_of_memory_region) >= len(cores_of_package):
-            cores_of_unit = cores_of_memory_region
-            logging.debug("Using memory regions as the basis for cpu core division")
-        else:
-            cores_of_unit = cores_of_package
-            logging.debug("Using physical packages as the basis for cpu core division")
-
-        # read hyper-threading information (sibling cores sharing the same physical core)
-        siblings_of_core = {}
-        for core in allCpus:
-            siblings = util.parse_int_list(
-                util.read_file(
-                    f"/sys/devices/system/cpu/cpu{core}/topology/thread_siblings_list"
-                )
-            )
-            siblings_of_core[core] = siblings
-        logging.debug("Siblings of cores are %s.", siblings_of_core)
+        # read list of available CPU cores (int)
+        allCpus_list = get_cpu_list(my_cgroups, coreSet)
+
+        # read & prepare hyper-threading information, filter superfluous entries
+        siblings_of_core = get_siblings_mapping (allCpus_list)
+        cleanList = []
+        for core in siblings_of_core:
+            if core not in cleanList:
+                for sibling in siblings_of_core[core]:
+                    if sibling != core:
+                        cleanList.append(sibling)
+        for element in cleanList:
+            siblings_of_core.pop(element)
+        # siblings_of_core will be added to hierarchy_levels list after sorting
+
+        # read & prepare mapping of cores to L3 cache
+        # cores_of_L3cache = 
+        # hierarchy_levels.append(core_of_L3cache)
+        
+        # read & prepare mapping of cores to NUMA region
+        cores_of_NUMA_Region = get_NUMA_mapping (allCpus_list)
+        hierarchy_levels.append(cores_of_NUMA_Region)
+
+        # read & prepare mapping of cores to group
+        # core_of_group =
+        #hierarchy_levels.append(core_of_group)
+
+        # read & prepare mapping of cores to CPU/physical package/socket?
+        cores_of_package = get_package_mapping (allCpus_list)
+        hierarchy_levels.append(cores_of_package)
+    
     except ValueError as e:
         sys.exit(f"Could not read CPU information from kernel: {e}")
-    return _get_cpu_cores_per_run0(
-        coreLimit,
-        num_of_threads,
-        use_hyperthreading,
+    
+    # generate sorted list of dicts in accordance with their hierarchy
+    def compare_hierarchy(dict1, dict2):
+        value1 = len(next(iter(dict1.values())))
+        value2 = len(next(iter(dict2.values())))
+        if value1 > value2:
+            return 1
+        if value1 < value2:
+            return -1
+        if value1 == value2:
+            return 0
+
+    # sort hierarchy_levels according to the dicts' corresponding unit sizes
+    hierarchy_levels.sort(key = cmp_to_key(compare_hierarchy))  #hierarchy_level = [dict1, dict2, dict3]
```

Comment:
An easier way to write this is to avoid `cmp_to_key` and just provide a function that returns an appropriate sorting key. Right now, `compare_hierarchy` looks up what should be compared (then lengths), then compares them, and then `cmp_to_key` undoes the comparison in order to provide some sort keys that `.sort()` can use. Just directly tell `.sort()` what it should compare on (and make sure to use `reverse` if necessary. Note that if the docs of `cmp_to_key` tell you to not use it in new code.